core/src/test/java/com/google/zxing/pdf417/decoder/PDF417DecoderTestCase.java

```diff
@@ -175,4 +181,197 @@ public void testSampleWithNoDataNoMacro() throws FormatException {
+  }
+
+  public void testUppercaseLowercaseNumericMix() throws WriterException, FormatException {
+    performPermutationTest(new char[] {'A','a','1'},7,15491);
+  }
+
+  @Test
+  public void testUppercaseLowercasePunctuationMix() throws WriterException, FormatException {
+    performPermutationTest(new char[] {'A','a',';'},7,15491);
+  }
+
+  @Test
+  public void testUppercaseLowercaseByteMix() throws WriterException, FormatException {
+    performPermutationTest(new char[] {'A','a','\u00c4'},7,17288);
+  }
+
+  @Test
+  public void testLowercasePunctuationByteMix() throws WriterException, FormatException {
+    performPermutationTest(new char[] {'a',';','\u00c4'},7,17427);
+  }
+
+  @Test
+  public void testUppercaseLowercaseNumericPunctuationMix() throws WriterException, FormatException {
+    performPermutationTest(new char[] {'A','a','1',';'},7,120479);
+  }
+
+  @Test
+  public void testBinaryData() throws WriterException, FormatException {
+    byte[] bytes = new byte[500];
+    Random random = new Random(0);
+    int total = 0;
+    for (int i = 0; i < 10000; i++) {
+      random.nextBytes(bytes);
+      total += encodeDecode(new String(bytes, StandardCharsets.ISO_8859_1));
+    }
+    assertEquals(4190044,total); 
+  }
+
+  private static void encodeDecode(String input,int expectedLength) throws WriterException, FormatException {
+    assertEquals(expectedLength, encodeDecode(input));
+  }
+
+  private static int encodeDecode(String input) throws WriterException, FormatException {
+    String s = PDF417HighLevelEncoderTestAdapter.encodeHighLevel(input, Compaction.AUTO, null);
+    int[] codewords = new int[s.length() + 1];
+    codewords[0] = codewords.length;
+    for (int i = 1; i < codewords.length; i++) {
+      codewords[i] = s.charAt(i - 1);
+    }
+    DecoderResult result = DecodedBitStreamParser.decode(codewords, "0");
+
+    assertEquals(input, result.getText());
+    return codewords.length;
+  }
+
+  private static int getEndIndex(int length, char[] chars) {
+    double decimalLength = Math.log(chars.length) / Math.log(10);
+    return (int) Math.ceil(Math.pow(10, decimalLength * length));
+  }
+
+  private static String generatePermutation(int index, int length, char[] chars) {
+    int N = chars.length;
+    String baseNNumber = Integer.toString(index, N);
+    while (baseNNumber.length() < length) {
+      baseNNumber = "0" + baseNNumber;
+    }
+    String prefix = "";
+    for (int i = 0; i < baseNNumber.length(); i++) {
+      prefix += chars[baseNNumber.charAt(i) - '0'];
+    }
+    return prefix;
+  }
+
+  private static void performPermutationTest(char[] chars,int length,int expectedTotal) throws WriterException,
+      FormatException {
+    int endIndex = getEndIndex(length, chars);
+    int total = 0;
+    for (int i = 0; i < endIndex; i++) {
+      total += encodeDecode(generatePermutation(i, length, chars));
+    }
+    String s = "['" + chars[0] + "'";
```

Comment:
If this is at all perf-sensitive, it can use StringBuilders instead, but may not matter for a test